corehq/apps/userreports/tasks.py

```diff
@@ -202,49 +203,73 @@ def queue_async_indicators():
 
-def _queue_indicator(redis_client, indicator):
-    lock_key = _get_indicator_queued_lock_key(indicator)
-    lock = redis_client.lock(lock_key, timeout=60 * 60 * 24)
-    if not lock.acquire(blocking=False):
-        return
 
-    indicator.date_queued = datetime.utcnow()
-    indicator.save()
-    save_document.delay(indicator.id, indicator.doc_id)
+def _queue_indicators(indicators):
+    def _queue_chunk(indicators):
+        now = datetime.utcnow()
 
+        # batch saves to one query
+        with transaction.atomic():
+            for indicator in indicators:
+                indicator.date_queued = now
+                indicator.save()
 
-def _get_indicator_queued_lock_key(indicator):
-    return 'async_indicator_queued-{}'.format(indicator.id)
+        indicator_doc_ids = [i.doc_id for i in indicators]
+        save_document.delay(indicator_doc_ids)
+
+    to_queue = []
+    for indicator in indicators:
+        to_queue.append(indicator)
+        if len(to_queue) >= ASYNC_INDICATOR_CHUNK_SIZE:
+            _queue_chunk(to_queue)
+            to_queue = []
+
+    _queue_chunk(to_queue)
+
+
+@quickcache(['config_id'])
+def _get_config(config_id):
+    # performance optimization for save_document. don't use elsewhere
+    return _get_config_by_id(config_id)
 
 
 @task(queue=UCR_INDICATOR_CELERY_QUEUE, ignore_result=True, acks_late=True)
-def save_document(async_indicator_id, doc_id):
-    lock_key = get_async_indicator_modify_lock_key(doc_id)
-    with CriticalSection([lock_key]):
-        indicator = AsyncIndicator.objects.get(pk=async_indicator_id)
-        doc_store = get_document_store(indicator.domain, indicator.doc_type)
-        doc = doc_store.get_document(indicator.doc_id)
-
-        eval_context = EvaluationContext(doc)
-        for config_id in indicator.indicator_config_ids:
-            config = _get_config_by_id(config_id)
-            adapter = get_indicator_adapter(config, can_handle_laboratory=True)
-            adapter.best_effort_save(doc, eval_context)
-            eval_context.reset_iteration()
-
-        redis_client = get_redis_client().client.get_client()
-        queued_lock_key = _get_indicator_queued_lock_key(indicator)
-        lock = redis_client.lock(queued_lock_key)
-        release_lock(lock, degrade_gracefully=True)
-        indicator.delete()
+def save_document(doc_ids):
+    lock_keys = []
+    for doc_id in doc_ids:
+        lock_keys.append(get_async_indicator_modify_lock_key(doc_id))
+
+    with CriticalSection(lock_keys):
+        indicators = AsyncIndicator.objects.filter(doc_id__in=doc_ids)
+        first_indicator = indicators[0]
+
+        for i in indicators:
+            assert i.domain == first_indicator.domain
+            assert i.doc_type == first_indicator.doc_type
+
+        doc_store = get_document_store(first_indicator.domain, first_indicator.doc_type)
+        for doc in doc_store.iter_documents(doc_ids):
+            indicator = next(i for i in indicators if doc['_id'] == i.doc_id)
```

Comment:
Instead of searching indicators for every doc, would it be quicker to create a dict of indicators keyed on doc_id?